mqtt/mqtt-bridge/src/rpc.rs

```diff
@@ -0,0 +1,503 @@
+            .await
+            .map_err(|e| RpcError::SendAck(command_id, e))
+    }
+}
+
+#[async_trait]
+impl EventHandler for RpcHandler {
+    type Error = RpcError;
+
+    async fn handle(&mut self, event: &Event) -> Result<(), Self::Error> {
+        match event {
+            Event::Publication(publication) => {
+                if let Some(command_id) = capture_command_id(&publication.topic_name) {
+                    let doc = Document::from_reader(&mut publication.payload.clone().reader())?;
+                    match bson::from_document(doc)? {
+                        VersionedRpcCommand::V1(command) => {
+                            self.handle_command(command_id, command).await?;
+                        }
+                    }
+                }
+            }
+            Event::SubscriptionUpdates(subscriptions) => {
+                for subscription in subscriptions {
+                    self.handle_subcription_update(subscription).await?;
+                }
+            }
+            _ => {}
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum RpcError {
+    #[error("failed to deserialize command from received publication")]
+    DeserializeCommand(#[from] bson::de::Error),
+
+    #[error("failed to serialize ack")]
+    SerializeAck(#[from] bson::ser::Error),
+
+    #[error("unable to send nack for {0}. {1}")]
+    SendNack(String, #[source] PublishError),
+
+    #[error("unable to send ack for {0}. {1}")]
+    SendAck(String, #[source] PublishError),
+}
+
+fn capture_command_id(topic_name: &str) -> Option<String> {
+    lazy_static! {
+        static ref RPC_TOPIC_PATTERN: Regex = Regex::new("\\$edgehub/rpc/(?P<command_id>[^/ ]+)$")
+            .expect("failed to create new Regex from pattern");
+    }
+
+    RPC_TOPIC_PATTERN
+        .captures(topic_name)
+        .and_then(|captures| captures.name("command_id"))
+        .map(|command_id| command_id.as_str().into())
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", tag = "cmd")]
+enum RpcCommand {
+    #[serde(rename = "sub")]
+    Subscribe {
+        #[serde(rename = "topic")]
+        topic_filter: String,
+    },
+
+    #[serde(rename = "unsub")]
+    Unsubscribe {
+        #[serde(rename = "topic")]
+        topic_filter: String,
+    },
+
+    #[serde(rename = "pub")]
+    Publish {
+        topic: String,
+
+        #[serde(with = "serde_bytes")]
+        payload: Vec<u8>,
+    },
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", tag = "version")]
+enum VersionedRpcCommand {
+    V1(RpcCommand),
+}
+
+#[cfg(test)]
+mod tests {
+    use bson::{bson, doc, spec::BinarySubtype};
+    use matches::assert_matches;
+    use mqtt3::{
+        proto::QoS, Event, MockPublishHandle, MockUpdateSubscriptionHandle, ReceivedPublication,
+    };
+    use test_case::test_case;
+
+    use super::*;
+
+    #[test]
+    fn it_deserizes_from_bson() {
```

Comment:
```suggestion
    fn it_deserializes_from_bson() {
```

Reply:
Also, what is the purpose of this test? it seems to me that you're just testing bson crate :)

Reply:
It verifies, that I assigned all serde features correct :) and yes, bson treats binary fields with some hiccups. see `serde_bytes` crate